proto-lens/src/Data/ProtoLens/Compare.hs

```diff
@@ -0,0 +1,108 @@
+-- Copyright 2016 Google Inc. All Rights Reserved.
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+-- | Matchers for comparing protocol buffer messages.
+module Data.ProtoLens.Compare (
+    -- * Matchers
+    partial,
+    ) where
+
+import Data.Constraint (Dict (..))
+import qualified Data.Map as Map
+import Lens.Family2 (Lens', over, set, view)
+import Lens.Family2.Unchecked (lens)
+import Data.ProtoLens.Message
+
+-- | Perform a partial match on a protobuffer. This corresponds to equality for
+-- primitive fields, (recursive) partial equality for required 'Message'
+-- fields, partial equality if both are set for optional fields, and partial
+-- equality of subsets for repeated and map fields.
+partial
+    :: forall a. Message a
+    => a  -- ^ The value to partially match against.
+    -> a  -- ^ The value to match.
+    -> Bool
+partial against match =
+  and . flip fmap (Map.elems (fieldsByTag $ descriptor @a))
+      $ \(FieldDescriptor _ ftd acc) ->
+      case acc of
+        PlainField _    l ->
+          eq ftd (view l against) (view l match)
+
+        OptionalField   l ->
+          compareMaybe (eq ftd) (view l against) (view l match)
+
+        RepeatedField _ l ->
+          and . fmap (\a -> any (eq ftd a) $ view l match) $ view l against
+
+        MapField k v l    ->
+          let entriesMatch   = toRepeatedEntry k v $ view l match
+              entriesAgainst = toRepeatedEntry k v $ view l against
+           in and $ fmap (\a -> any (eq ftd a) entriesMatch) entriesAgainst
+  where
+    toRepeatedEntry :: (Ord key, Message entry)
```

Comment:
Break these helper functions out into separate, top-level (unexported) functions.

Ok if it makes sense to group them together; for example, make `getEq` part of the where clause of `eq`.